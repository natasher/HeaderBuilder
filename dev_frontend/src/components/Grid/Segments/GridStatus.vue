<template>
  <div class = "status">
    <span class = "label">Status</span>

    <ul>

      <li
        v-show = "getStickyFlag === 'sticky'"
        :class = "{ 'active': getGridStatus === 'off' }"
        @click.passive = "setGridStatus({ status: 'off' })" >
          Off
      </li>

      <li
        :class = "{ 'active': getGridStatus === 'custom' }"
        @click.passive = "setGridStatus({ status: 'custom' })" >
          Custom
      </li>

      <li
        v-show = "getCurrentView !== 'desktop' || getStickyFlag !== 'default'"
        :class = "{ 'active': getGridStatus === 'auto' }"
        @click.passive = "handleAutoClick({ $modal: $modal })" >
          Auto
      </li>

    </ul>
  </div>
</template>

<script>
import { mapActions, mapGetters, mapState } from 'vuex'

export default {

  name: 'GridStatus',

  methods: mapActions( 'ui', [
            'setGridStatus',
            'handleAutoClick'
          ]),

  computed: {
    ...mapState( 'general', {
      getCurrentView: state => state.currentView,
    }),

    ...mapState( 'ui', {
      getStickyFlag : state => state.stickyFlag,
    }),

    ...mapGetters( 'ui', [ 'getGridStatus' ]),
  },

}
</script>
