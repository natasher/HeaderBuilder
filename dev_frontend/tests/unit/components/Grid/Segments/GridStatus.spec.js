import { shallowMount, createLocalVue } from '@vue/test-utils'
import Vuex          from 'vuex'
import flushPromises from 'flush-promises'
import merge         from 'lodash.merge'
import GridStatus    from '../../../../../src/components/Grid/Segments/GridStatus.vue'

const localVue = createLocalVue()
localVue.use( Vuex )

function createStore(overrides) {
  const defaultStoreConfig = {
    modules: {
      general: {
        namespaced: true,
        state: {
          currentView: 'desktop',
        }
      },
      ui: {
        namespaced: true,
        state: {
          stickyFlag: 'sticky',
        },
        actions: {
          setGridStatus: jest.fn(),
          handleAutoClick: jest.fn(),
        },
        getters: {
          getGridStatus: jest.fn(() => { return 'custom' })
        },
      }
    }
  }

  return new Vuex.Store( merge( defaultStoreConfig, overrides ))
}

function createWrapper(overrides) {
  const defaultMountingOptions = {
    localVue,
    store: createStore(),
  }

  return shallowMount( GridStatus, merge( defaultMountingOptions, overrides ))
}

describe('GridStatus.vue', () => {

  test('should match snapshot if grid status is `custom` and header is `sticky`, device `mobile`', () => {
    const store = createStore({
      modules: {
        general: {
          state: {
            currentView: 'mobile',
          }
        },
        ui: {
          state: {
            stickyFlag: 'sticky',
          },
          getters: {
            getGridStatus: jest.fn(() => { return 'custom' })
          },
        }
      }
    })
    const wrapper = createWrapper({ store, })

    expect( wrapper.element ).toMatchSnapshot()
  })

  test('should show `Off` and `Custom` tab if header is `sticky`', () => {
    const store = createStore({
      modules: {
        general: {
          state: {
            currentView: 'tablet',
          }
        },
        ui: {
          state: {
            stickyFlag: 'sticky',
          },
          getters: {
            getGridStatus: jest.fn(() => { return 'off' })
          },
        }
      }
    })
    const wrapper = createWrapper({ store, })
    const tabs = wrapper.findAll('li')

    expect( tabs.at(0).isVisible() ).toBe( true )
    expect( tabs.at(0).classes() ).toContain( 'active' )

    expect( tabs.at(1).isVisible() ).toBe( true )
    expect( tabs.at(1).classes() ).not.toContain( 'active' )

    expect( tabs.at(2).isVisible() ).toBe( true )
    expect( tabs.at(2).classes() ).not.toContain( 'active' )
  })

  test('should NOT show `Off` and show `Custom` tab if header is different then `sticky`', () => {
    const store = createStore({
      modules: {
        general: {
          state: {
            currentView: 'tablet',
          }
        },
        ui: {
          state: {
            stickyFlag: 'default',
          },
          getters: {
            getGridStatus: jest.fn(() => { return 'custom' })
          },
        }
      }
    })
    const wrapper = createWrapper({ store, })
    const tabs = wrapper.findAll('li')

    expect( tabs.at(0).isVisible() ).toBe( false )
    expect( tabs.at(0).classes() ).not.toContain( 'active' )

    expect( tabs.at(1).isVisible() ).toBe( true )
    expect( tabs.at(1).classes() ).toContain( 'active' )

    expect( tabs.at(2).isVisible() ).toBe( true )
    expect( tabs.at(2).classes() ).not.toContain( 'active' )
  })

  test('should NOT show `Auto` tab if header is `desktop` and `default`', () => {
    const store = createStore({
      modules: {
        general: {
          state: {
            currentView: 'desktop',
          }
        },
        ui: {
          state: {
            stickyFlag: 'default',
          },
          getters: {
            getGridStatus: jest.fn(() => { return 'custom' })
          },
        }
      }
    })
    const wrapper = createWrapper({ store, })
    const tabs = wrapper.findAll('li')

    expect( tabs.at(0).isVisible() ).toBe( false )
    expect( tabs.at(0).classes() ).not.toContain( 'active' )

    expect( tabs.at(1).isVisible() ).toBe( true )
    expect( tabs.at(1).classes() ).toContain( 'active' )

    expect( tabs.at(2).isVisible() ).toBe( false )
    expect( tabs.at(2).classes() ).not.toContain( 'active' )
  })

  test('should show `Auto` if header is different then `desktop` and `default`', () => {
    const store = createStore({
      modules: {
        general: {
          state: {
            currentView: 'tablet',
          }
        },
        ui: {
          state: {
            stickyFlag: 'default',
          },
          getters: {
            getGridStatus: jest.fn(() => { return 'custom' })
          },
        }
      }
    })
    const wrapper = createWrapper({ store, })
    const tabs = wrapper.findAll('li')

    expect( tabs.at(0).isVisible() ).toBe( false )
    expect( tabs.at(0).classes() ).not.toContain( 'active' )

    expect( tabs.at(1).isVisible() ).toBe( true )
    expect( tabs.at(1).classes() ).toContain( 'active' )

    expect( tabs.at(2).isVisible() ).toBe( true )
    expect( tabs.at(2).classes() ).not.toContain( 'active' )
  })

  test('should change grid status to `off` if `Off` tab is clicked', async () => {
    const setGridStatus = jest.fn()
    const store = createStore({
      modules: {
        ui: {
          actions: {
            setGridStatus,
          },
        }
      }
    })
    const wrapper = createWrapper({ store, })
    const offTab  = wrapper.findAll('li').at( 0 )

    offTab.trigger( 'click' )
    await flushPromises()

    expect(setGridStatus).toBeCalledWith(
      expect.anything(),
      { status: 'off' },
      void 0
    )
  })

  test('should change grid status to `off` if `Off` tab is clicked', async () => {
    const handleAutoClick = jest.fn()
    const $modal = jest.fn()
    const store = createStore({
      modules: {
        ui: {
          actions: {
            handleAutoClick,
          },
        }
      }
    })
    const wrapper = createWrapper({
      store,
      mocks: {
        $modal,
      }
    })
    const autoTab = wrapper.findAll('li').at( 2 )

    autoTab.trigger( 'click' )
    await flushPromises()

    expect(handleAutoClick).toBeCalledWith(
      expect.anything(),
      { $modal: $modal, },
      void 0
    )
  })
})