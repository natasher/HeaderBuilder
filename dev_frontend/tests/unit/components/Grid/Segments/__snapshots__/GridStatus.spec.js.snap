// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`GridStatus.vue should match snapshot if grid status is \`custom\` and header is \`sticky\`, device \`mobile\` 1`] = `
<div
  class="status"
>
  <span
    class="label"
  >
    Status
  </span>
   
  <ul>
    <li
      class=""
    >
      
        Off
    
    </li>
     
    <li
      class="active"
    >
      
        Custom
    
    </li>
     
    <li
      class=""
    >
      
        Auto
    
    </li>
  </ul>
</div>
`;
